ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/TransactionBroadcaster.java

```diff
@@ -0,0 +1,144 @@
+
+public class TransactionBroadcaster implements TransactionBatchAddedListener {
+
+  private final AbstractPendingTransactionsSorter pendingTransactions;
+  private final PeerTransactionTracker transactionTracker;
+  private final TransactionsMessageSender transactionsMessageSender;
+  private final NewPooledTransactionHashesMessageSender newPooledTransactionHashesMessageSender;
+  private final EthContext ethContext;
+  private final int numPeersToSendFullTransactions;
+
+  public TransactionBroadcaster(
+      final EthContext ethContext,
+      final AbstractPendingTransactionsSorter pendingTransactions,
+      final PeerTransactionTracker transactionTracker,
+      final TransactionsMessageSender transactionsMessageSender,
+      final NewPooledTransactionHashesMessageSender newPooledTransactionHashesMessageSender) {
+    this.pendingTransactions = pendingTransactions;
+    this.transactionTracker = transactionTracker;
+    this.transactionsMessageSender = transactionsMessageSender;
+    this.newPooledTransactionHashesMessageSender = newPooledTransactionHashesMessageSender;
+    this.ethContext = ethContext;
+    this.numPeersToSendFullTransactions =
+        (int) Math.ceil(Math.sqrt(ethContext.getEthPeers().getMaxPeers()));
+  }
+
+  public void relayTransactionPoolTo(final EthPeer peer) {
+    Set<TransactionInfo> pendingTransactionInfo = pendingTransactions.getTransactionInfo();
+    if (!pendingTransactionInfo.isEmpty()) {
+      if (peer.hasSupportForMessage(EthPV65.NEW_POOLED_TRANSACTION_HASHES)) {
+        sendTransactionHashes(toTransactionList(pendingTransactionInfo), List.of(peer));
+      } else {
+        sendFullTransactions(toTransactionList(pendingTransactionInfo), List.of(peer));
+      }
+    }
+  }
+
+  @Override
+  public void onTransactionsAdded(final Iterable<Transaction> transactions) {
+    final int currPeerCount = ethContext.getEthPeers().peerCount();
+    if (currPeerCount == 0) {
+      return;
+    }
+
+    List<EthPeer> peersWithOnlyTransactionSupport = new ArrayList<>(currPeerCount);
+    List<EthPeer> peersWithTransactionHashesSupport = new ArrayList<>(currPeerCount);
+
+    ethContext
+        .getEthPeers()
+        .streamAvailablePeers()
+        .forEach(
+            peer -> {
+              if (peer.hasSupportForMessage(EthPV65.NEW_POOLED_TRANSACTION_HASHES)) {
+                peersWithTransactionHashesSupport.add(peer);
+              } else {
+                peersWithOnlyTransactionSupport.add(peer);
+              }
+            });
+
+    if (peersWithOnlyTransactionSupport.size() < numPeersToSendFullTransactions) {
+      final int delta =
+          Math.min(
+              numPeersToSendFullTransactions - peersWithOnlyTransactionSupport.size(),
+              peersWithTransactionHashesSupport.size());
+      // add peer from the other list to reach the required size
+      Collections.shuffle(peersWithTransactionHashesSupport);
+      IntStream.range(0, delta)
```

Comment:
Maybe I don't understand properly what is going on here, I feel this would be more understandable if we would just use something like `peersWithOnlyTransactionSupport.addAll(peersWithTransactionHashesSupport.sublist(..)); 
peersWithTransactionHashesSupport.removeRange(...);`
This way it looks like a for cycle obscured in a functional way.

Reply:
yes it could be hard to read, I like your suggestion but `removeRange` is protected, so I am rewriting this with a plain for cycle